EDU 334: update List Workflow information (#2115)
temporalio/documentation: docs-src/go/how-to-list-workflow-executions-using-the-client-in-go.md

  - client
---

The `ListWorkflow()` function retrieves [Workflow Executions](/concepts/what-is-a-workflow-execution) based on a [Query](/concepts/what-is-a-query) or [List Filter](/concepts/what-is-a-list-filter).

Use a Query to retrieve Workflow Executions based on state characteristics.
Queries include one or more conditions to return Workflow Executions with relevant [Workflow Ids](/concepts/what-is-a-workflow-id), [Workflow Types](/concepts/what-is-a-workflow-type), or start and end times.

Use a List Filter to retrieve Workflow Executions that contain certain [Search Attribute](/concepts/what-is-a-search-attribute) characteristics, such as name, values, and operators.
List Filters return a list of valid Workflow Executions from the Visibility store.

To list all Workflow Executions, use [`ListWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) on the [Temporal Client](/concepts/what-is-a-temporal-client).

```go
resp, err := temporalClient.ListWorkflow(ctx.Background(), req)
if err != nil {
  return err
}

fmt.Println("First page of results:")
for _, exec := range resp.Executions {
  fmt.Println("Workflow ID %v\n", exec.Execution.WorkflowId)

>   fmt.Println("Next page of results:")

Author: I'm going to take out the paging and just have it as a list.

  - client
---

> The [`ListWorkflow()`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) function retrieves a list of [Workflow Executions](/concepts/what-is-a-workflow-execution) that match the [Search Attributes](/concepts/what-is-a-search-attribute) of a given [List Filter](/concepts/what-is-a-list-filter).

Reviewer: ```suggestion
The [ListWorkflow()](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) function retrieves a list of [Workflow Executions](/concepts/what-is-a-workflow-execution) that match the [Search Attributes](/concepts/what-is-a-search-attribute) of a given [List Filter](/concepts/what-is-a-list-filter).
```

> The metadata returned from the [Visibility store](/concepts/what-is-visibility) can be used to get a Workflow Execution's history and details from the [Persistence store](/concepts/what-is-a-temporal-cluster#persistence).

Reviewer: What do you think about the following changes to link text to more closely match the text to the target?
```suggestion
The metadata returned from the [Visibility](/concepts/what-is-visibility) store can be used to get a Workflow Execution's history and details from the [Persistence](/concepts/what-is-a-temporal-cluster#persistence) store.
```

Use a List Filter to define a `request` to pass into `ListWorkflow()`.

```go

> request := "CloseTime = missing"

Reviewer 2: `request` has to be a full `ListWorkflowExecutionsRequest` object here, not just a string, so something like:

```suggestion
request := &workflowservice.ListWorkflowExecutionsRequest{ Query: "CloseTime = missing" }
```

```

This `request` value returns only open Workflows.

> For more List Filter examples, see the [examples provided for List Filters in Temporal's Visibility guide.](/concepts/what-is-a-list-filter#list-filter-examples)

Reviewer: ```suggestion
For more List Filter examples, see the [examples provided for List Filters in the Temporal Visibility guide.](/concepts/what-is-a-list-filter#list-filter-examples)
```

```go
resp, err := temporalClient.ListWorkflow(ctx.Background(), request)
if err != nil {
  return err
}

> fmt.Println("List of results:")

Reviewer 2: ```suggestion
fmt.Println("First page of results:")
```

I liked that you made it clear before that this only the first page of results. If users want more than the first page, they have to do extra work (they have to re-invoke).

for _, exec := range resp.Executions {

>   fmt.Println("Workflow ID %v\n", exec.Execution.WorkflowId)

Reviewer 2: ```suggestion
  fmt.Println("Workflow ID %v", exec.Execution.WorkflowId)
```

  - client
---

The `ListWorkflow()` function retrieves [Workflow Executions](/concepts/what-is-a-workflow-execution) that match the properties specified in a [List Filter](/concepts/what-is-a-list-filter).

> These properties are defined by [Search Attributes](/concepts/what-is-a-search-attribute), which are used to retrieve Workflow Executions from the [Visibility store](/concepts/what-is-visibility).

Reviewer: Should `Visibility store` link to `/cluster-deployment-guide#visibility-store`?

Reviewer 3: I think the specificity is a little bit off...

WFEs are not stored in the visibility store - they are stored in the normal persistence. Search attribute data is stored in the visibility store (what search attributes are part of the system basically)

WFEs are stored in normal persistence.
Search Attribute source of truth data is in the Vis store.
The ListWorkflow function is used to retrieve a list of WFs
The List Filter provided to the ListWorkflow call filters and sorts the list.
What is returned is WFE meta data - 
You can use that to get the full details and history of and individual WFE.

Use a List Filter to define a `request` to pass into `ListWorkflow()`.

```go
// ...
request := "CloseTime = missing"
```

> This `request` value will only return open Workflows.

Reviewer: ```suggestion
This `request` value returns only open Workflows.
```

> This `request` value will only return open Workflows.

Reviewer 3: More List Filter syntax examples can be found... and link to that spot?

> To list all Workflow Executions that satisfy this List Filter, use [`ListWorkflow()`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) on the [Temporal Client](/concepts/what-is-a-temporal-client).

Reviewer: ```suggestion
To list all Workflow Executions that satisfy this List Filter, use [ListWorkflow()](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) on the [Temporal Client](/concepts/what-is-a-temporal-client).
```

Reviewer 3: This is a bit verbose and repetitive -
You could wipe this whole sentence if the API definition is linked in the first mention of the ListWorkflow API in the first sentence.

```go
// ...

resp, err := temporalClient.ListWorkflow(ctx.Background(), request)
if err != nil {
  return err
}

fmt.Println("List of results:")
for _, exec := range resp.Executions {
  fmt.Println("Workflow ID %v\n", exec.Execution.WorkflowId)
}

> //...

Reviewer: ```suggestion
// ...
```
Or remove the space from the first line. I don't think we've defined a style yet, so let's just aim for consistency.

  - client
---

There are several functions for listing [Workflow Executions](/concepts/what-is-a-workflow-execution) available on the [Temporal Go SDK](https://pkg.go.dev/go.temporal.io/sdk).
Each function retrieves Workflow Executions according to a given [Query](/concepts/what-is-a-query) or request filters.

To list all Workflow Executions, use [`ListWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) on the [Temporal Client](/concepts/what-is-a-temporal-client).

```go
resp, err := temporalClient.ListWorkflow(context.Background(), sampleListRequest)

> // print?

Reviewer 3: do we need these print comments?

Author: Still trying to figure out if I need to include a print statement for the response or if it just does it of its own accord.

Author: I wasn't sure, so I'm gonna ask Roey or somebody

  - client
---

> There are several functions for listing [Workflow Executions](/concepts/what-is-a-workflow-execution) available on the [Temporal Go SDK](https://pkg.go.dev/go.temporal.io/sdk).
> Each function retrieves Workflow Executions according to a given [Query](/concepts/what-is-a-query) or request filters.

Reviewer: ```suggestion
Several functions for listing [Workflow Executions](/concepts/what-is-a-workflow-execution) are available on the [Temporal Go SDK](https://pkg.go.dev/go.temporal.io/sdk).
Each function retrieves Workflow Executions according to a specified [Query](/concepts/what-is-a-query) or request filters.
```

To list all Workflow Executions, use [`ListWorkflow`](https://pkg.go.dev/go.temporal.io/sdk/client#Client.ListWorkflow) on the [Temporal Client](/concepts/what-is-a-temporal-client).

```go
resp, err := temporalClient.ListWorkflow(ctx.Background(), req)
if err != nil {
  return err
}

fmt.Println("First page of results:")
for _, exec := range resp.Executions {
  fmt.Println("Workflow ID %v\n", exec.Execution.WorkflowId)
  fmt.Println("Next page of results:")
}
```

To return a list of archived Workflow Executions, use `ListArchivedWorkflow()` on the Client.

> Make sure that the Temporal Cluster or target [Namespace](/concepts/what-is-a-namespace) is configured for [visibility archival](/concepts/what-is-standard-visibility) before using this function.

Reviewer: ```suggestion
Make sure that the Temporal Cluster or target [Namespace](/concepts/what-is-a-namespace) is configured for [Archival](/concepts/what-is-archival) before using this function.
```
Is this change accurate? (I could also make a case for linking to https://temporal-documentation-ao26cn23g.preview.thundergun.io/cluster-deployment-guide#set-up-archival, but that information is only one click away from the definition of Archival.)
